### PR TITLE
Implement Attach

### LINKED ISSUE #10: Support attach to Pod

### PR DESCRIPTION
**Pull request by brendandburns**
(also re-factor the Exec library)

Fixes #10

### CODE REVIEW

**Comment by mbohlool** [comment]
This needs rebase to remove exec commit.

File: examples/src/main/java/io/kubernetes/client/examples/AttachExample.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+import java.io.BufferedReader;
+import java.io.InputStreamReader;
+import java.io.IOException;
+import java.io.OutputStream;
+
+/**
+ * A simple example of how to use the Java API
+ * 
+ * Requires kubectl proxy running
```
Comment (mbohlool) [inline examples/src/main/java/io/kubernetes/client/examples/AttachExample.java]: I don't think we need this anymore.

File: util/src/main/java/io/kubernetes/client/Attach.java [hunk 2]
```diff
@@ -0,0 +1,161 @@
+package io.kubernetes.client;
```
Comment (mbohlool) [inline util/src/main/java/io/kubernetes/client/Attach.java]: Header?

File: util/src/main/java/io/kubernetes/client/Exec.java [hunk 3]
```diff
@@ -1,254 +1,256 @@
...
-                ex.printStackTrace();
-            }
-            // TODO: get status code here
-            synchronized(this) {
-                this.notifyAll();
-            }
-        }
-    }
-}
+package io.kubernetes.client;
```
Comment (mbohlool) [inline util/src/main/java/io/kubernetes/client/Exec.java]: Header?

File: util/src/main/java/io/kubernetes/client/util/WebSocketStreamHandler.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+import java.io.InputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.io.PipedInputStream;
+import java.io.PipedOutputStream;
+import java.io.Reader;
+import java.util.HashMap;
+import java.util.Map;
+
+public class WebSocketStreamHandler implements WebSockets.SocketListener {
```
Comment (mbohlool) [inline util/src/main/java/io/kubernetes/client/util/WebSocketStreamHandler.java]: javadoc?

**Comment by mbohlool** [comment]
I did a fast review. need to look in more depth. What I am interested to look deeper for is, to my understanding, attach and exec are fairly similar operations but there are rather big differences between two classes implementing it. I would like to see if we can get them closer and even share some code between them. Meanwhile, I added some high-level comments. Feel free to ignore them until I look at this again.

**Comment by brendandburns** [comment]
My plan was to refactor Exec in terms of the `WebSocketHandler` class, once this merges, which should unify things somewhat more. If you want me to do that in this PR, I can do that.

File: util/src/main/java/io/kubernetes/client/Attach.java (hunk 2 above)
Comment (brendandburns) [inline util/src/main/java/io/kubernetes/client/Attach.java]: done.

File: util/src/main/java/io/kubernetes/client/Exec.java (hunk 3 above)
Comment (brendandburns) [inline util/src/main/java/io/kubernetes/client/Exec.java]: done.

File: util/src/main/java/io/kubernetes/client/util/WebSocketStreamHandler.java (hunk 4 above)
Comment (brendandburns) [inline util/src/main/java/io/kubernetes/client/util/WebSocketStreamHandler.java]: Done.

**Comment by brendandburns** [comment]
@mbohlool I went ahead and just did the refactor while I addressed your comments. Please take another look.

**Comment by brendandburns** [comment]
@mbohlool friendly ping on this, I think it is ready to merge...

File: util/src/main/java/io/kubernetes/client/util/WebSocketStreamHandler.java [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+    public void open(String protocol, Closeable closer) {
+        this.protocol = protocol;
+        this.closer = closer;
+    }
+        
+    @Override
+    public void bytesMessage(InputStream in) {
+        try {
+            OutputStream out = getOutputStream(in.read());
+            byte[] buffer = new byte[4096];
```
Comment (mbohlool) [inline util/src/main/java/io/kubernetes/client/util/WebSocketStreamHandler.java]: nit: a constant for 4096? (BufferSize)

File: util/src/main/java/io/kubernetes/client/util/WebSocketStreamHandler.java [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+            ex.printStackTrace();
+        }
+    }
+        
+    @Override
+    public void textMessage(Reader in) {
+        try {
+            // TODO: there has to be a better way to do this...
+            OutputStream out = getOutputStream(in.read());
+            char[] buffer = new char[4096];
```
Comment (mbohlool) [inline util/src/main/java/io/kubernetes/client/util/WebSocketStreamHandler.java]: nit: the constant can be used here!

File: util/src/main/java/io/kubernetes/client/util/WebSocketStreamHandler.java [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+        }
+        return input.get(stream);
+    }
+
+    /**
+     * Get a specific output stream using its identifier
+     * @param stream The stream to return
+     * @returns The specified stream.
+     */
+    public synchronized OutputStream getOutputStream(int stream) {
```
Comment (mbohlool) [inline util/src/main/java/io/kubernetes/client/util/WebSocketStreamHandler.java]: I tried to run Exec example and interactively run commands on a pod but I couldn't do it. The thing is I don't see anywhere in this code that actually sends command to WebSocket object. the WebSocket listener onOpen callback returns a websocket object that is completely ignored. The outputStreams in this code and the code before this PR both are connected to a piped input stream that does nothing with the received data (in my example my keystrokes). These should be somehow sent to websocket object's writeMessage method.

**Comment by mbohlool** [comment]
Sorry for the delay. you can always ignore nit comments. But I have a functional comment. I tried to interactively run a command on a pod using exec example and failed. I think the code lack the ability to send commands using OutputStream (not through exec's parameters).

File: util/src/main/java/io/kubernetes/client/util/WebSocketStreamHandler.java (hunk 5 above)
Comment (brendandburns) [inline util/src/main/java/io/kubernetes/client/util/WebSocketStreamHandler.java]: Removed in favor of `ByteStreams.copy(...)` from Guava.

File: util/src/main/java/io/kubernetes/client/util/WebSocketStreamHandler.java (hunk 7 above)
Comment (brendandburns) [inline util/src/main/java/io/kubernetes/client/util/WebSocketStreamHandler.java]: This is fixed now.  `stdin` was never really implemented before.   Now it is, please take another look.

**Comment by brendandburns** [comment]
proper input handling now added.  Please take a look when you get a chance.

**Comment by brendandburns** [comment]
@mbohlool rebased, please take another look when you get a chance.

Thanks!

File: examples/src/main/java/io/kubernetes/client/examples/ExecExample.java [hunk 8]
```diff
@@ -32,19 +42,15 @@ public static void main(String[] args) throws IOException, ApiException, Interru
         Configuration.setDefaultApiClient(client);
 
         Exec exec = new Exec();
-        final Process proc = exec.exec("default", "nginx-2371676037-czqx3", new String[] {"sh", "-c", "echo foo"}, true);
+        boolean tty = System.console() != null;
```
Comment (mbohlool) [inline examples/src/main/java/io/kubernetes/client/examples/ExecExample.java]: No need to fix in this PR, but I suggest we create the pod too so the example is self contained. here is the python example: https://github.com/kubernetes-incubator/client-python/blob/master/examples/exec.py

**Comment by mbohlool** [comment]
Looks good. Only one comment that can be addressed in another PR. I am going to merge this.

**Pull request merged**